EIPS/eip-5131.md

```diff
@@ -0,0 +1,241 @@
+            unchecked {
+              i--;
+            }
+        }
+        bytes32 ensCheck = keccak256(ensCheckBuffer);
+        
+        // Length of senderENS must be >= ensCheckBuffer.length+4 (since it needs to be of format auth[0-9]*.mainENS)
+        require(senderENS.length >= ensCheckBuffer.length+4, "Invalid");
+        // Check ending substring of the senderENS matches
+        require(ensCheck == keccak256(senderENS[senderENS.length-ensCheckBuffer.length:]), "Invalid");
+        // Check prefix matches auth[0-9]*.
+        require(keccak256(abi.encodePacked('auth')) == keccak256(senderENS[:4]), "Invalid");
+        for (uint i = senderENS.length-ensCheckBuffer.length; i > 4;) {
+          require(senderENS[0] >= 0x30 && senderENS[i] <= 0x39, "Invalid");
+          unchecked {
+            i--;
+          }
+        }
+        return true;
+    }
+
+    // *********************
+    //   Helper Functions
+    // *********************
+
+    function _computeNamehash(string[] memory _nameParts) private pure returns (bytes32 namehash) {
+        namehash = 0x0000000000000000000000000000000000000000000000000000000000000000;
+        for (uint i = _nameParts.length; i > 0;) {
+            namehash = keccak256(
+              abi.encodePacked(namehash, keccak256(abi.encodePacked(_nameParts[i-1])))
+            );
+            unchecked {
+              i--;
+            }
+        }
+    }
+
+    function _computeReverseNamehash() private view returns (bytes32 namehash) {
+        namehash = 0x0000000000000000000000000000000000000000000000000000000000000000;
+        namehash = keccak256(
+          abi.encodePacked(namehash, keccak256(abi.encodePacked('reverse')))
+        );
+        namehash = keccak256(
+          abi.encodePacked(namehash, keccak256(abi.encodePacked('addr')))
+        );
+        namehash = keccak256(
+          abi.encodePacked(namehash, keccak256(_addressToStringLowercase(msg.sender)))
+        );
+    }
+
+    function _addressToStringLowercase(address _address) private pure returns (bytes memory addressString) {
+        addressString = new bytes(40);
+        for (uint i = 0; i < 20; i++) {
+            bytes1 b = bytes1(uint8(uint(uint160(_address)) / (2**(8*(19 - i)))));
+            bytes1 hi = bytes1(uint8(b) / 16);
+            bytes1 lo = bytes1(uint8(b) - 16 * uint8(hi));
+            addressString[2*i] = _bytes1ToChar(hi);
+            addressString[2*i+1] = _bytes1ToChar(lo);            
+        }
+    }
+
+    function _bytes1ToChar(bytes1 b) private pure returns (bytes1 c) {
+        if (uint8(b) < 10) return bytes1(uint8(b) + 0x30);
+        else return bytes1(uint8(b) + 0x57);
+    }
+
+}
+```
+
+## Security Considerations
+The core purpose of this EIP is to enhance security and promote a safer way to authenticate wallet control and asset ownership when the main wallet is not needed and assets held by the main wallet do not need to be moved.  Consider it a way to do 'read only' authentication.
```

Comment:
I think that it's worth mentioning that when you register an `authAddress`, you need to be cognizant of the fact that this address can now be given access to various operations that it previously didn't have access to.

Reply:
As per our convo: 

If i set an authAddress any application can now accept that in leu of the mainAddress so as a user I want to be okay with that fact. Otherwise we could introduce another concept that allows you to narrow the scope to applications.
e.g. `authOpenSea.<mainENS>`, `authManifold.<mainENS>`

If every third party app piggybacks on auth.<mainENS> then the user can inadvertently expose a single hot wallet key to more power than they intended.

Maybe `auth[a-zA-Z0-9]*.<mainENS>` would be useful?

Reply:
I agree, I think `auth[a-zA-Z0-9]*` would be neat to have in general (and make the onchain pattern verification much simpler).

Reply:
+1 for `auth[a-zA-Z0-9]*`, it will allow for more clarity as well when potentially managing multiple authAddresses for them to actually be named

Reply:
interesting, but I'd argue that this is beyond the scope of this proposal. imo, it warrants an own eip since there are much more things to consider to standardise this.

anyways, I did not fully understand how this would work yet. would `mainENS` point to a standard ERC721 contract in this case? If so, it will be impossible to set the required reverse records. or did I understand something wrong here?

Reply:
Yeah agree its out of scope, but this EIP just got me thinking 😅 

> If so, it will be impossible to set the required reverse records
Would think the `owner` of the ERC721 contract would be `mainENS` in this case

Anyways, maybe I'll throw up a discussion somewhere else

Reply:
> Would think the owner of the ERC721 contract would be mainENS in this case

Yup, that could work. but it's outside the ERC721 specs. so it would only apply to a certain subset of collections that use the `Ownable` extension.
Definitely work discussing it though - please ping me if you spin something up.